src/wp-includes/class-wp-theme.php

```diff
@@ -1714,6 +1722,46 @@ public static function get_allowed_on_site( $blog_id = null ) {
 		return (array) apply_filters( 'site_allowed_themes', $allowed_themes[ $blog_id ], $blog_id );
 	}
 
+	/**
+	 * Returns the folder names of the block template directories.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @return string[] {
+	 *     Folder names used by block themes.
+	 *
+	 *     @type string $wp_template      Theme-relative directory name for block templates.
+	 *     @type string $wp_template_part Theme-relative directory name for block template parts.
+	 * }
+	 */
+	public function get_block_template_folders() {
+		if ( isset( $this->block_template_folders ) ) {
+			return $this->block_template_folders;
+		}
+
+		$paths_to_block_template_directories = array(
```

Comment:
Rather than create two, maybe three arrays, and run a `foreach()`, what about this?

Using `file_exists()` also means that this checks whether the directory is readable.

```php
if ( isset( $this->block_template_folders ) ) {
	return $this->block_template_folders;
}

$block_template_parts_path = $this->theme_root . '/' . $this->stylesheet . '/block-template-parts';
$block_templates_path      = $this->theme_root . '/' . $this->stylesheet . '/block-templates';

if ( file_exists( $block_template_parts_path ) || file_exists( $block_templates_path ) ) {
	$this->block_template_folders = array(
		'wp_template'      => 'block-templates',
		'wp_template_part' => 'block-template-parts',
	);
} else {
	$this->block_template_folders = array(
		'wp_template'      => 'templates',
		'wp_template_part' => 'parts',
	);
}

return $this->block_template_folders;
```

Reply:
@costdev I have updated the code to address this.